
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Methods &#8212; ChlamDB 2.0 (June 2019) documentation</title>
    <link rel="stylesheet" href="../_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/css/bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="../_static/css/custom.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/language_data.js"></script>
    <script type="text/javascript" src="../_static/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/js/custom.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Website tutorial" href="../tutorial/website.html" />
    <link rel="prev" title="ChlamDB documentation" href="../index.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../tutorial/website.html" title="Website tutorial"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="../index.html" title="ChlamDB documentation"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">ChlamDB 2.0 (June 2019) documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="methods">
<h1>Methods<a class="headerlink" href="#methods" title="Permalink to this headline">¶</a></h1>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>A simplified scheme of <code class="docutils literal notranslate"><span class="pre">ChlamDB</span></code> annotation workflow is shown in <strong>Figure 1</strong>. In summary:</p>
<blockquote>
<div><ul class="simple">
<li><p>Annotated genome assemblies integrated into ChlamDB were downloaded from <a class="reference external" href="https://www.ncbi.nlm.nih.gov/genbank/">GenBank</a> (or
<a class="reference external" href="https://www.ncbi.nlm.nih.gov/refseq/">RefSeq</a> if the GenBank assembly was not annotated).</p></li>
<li><p>Protein sequences were annotated based on data from muliple source databases and annotation softwares.</p></li>
<li><dl class="simple">
<dt>Protein sequences were clustered into orthologous groups (orthogroups) with <a class="reference external" href="https://github.com/davidemms/OrthoFinder">OrthoFinder</a>.</dt><dd><ul>
<li><p>An alignment and a phylogeny was reconstructed for each orthologous group.</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>The closest homologs of each protein were identified in <a class="reference external" href="https://www.ncbi.nlm.nih.gov/refseq/">RefSeq</a> and <a class="reference external" href="https://www.uniprot.org/">SwissProt</a></dt><dd><ul>
<li><p>A phylogeny including the closest non-PVC homologs was reconstructed for each orthogroup</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>Data were then integrated into a custom SQL database derived from the <a class="reference external" href="https://biosql.org/wiki/Main_Page">bioSQL</a> relational model.</p></li>
</ul>
</div></blockquote>
<div class="align-center figure align-default" id="id1">
<img alt="../_images/workflowV2.svg" src="../_images/workflowV2.svg" /><p class="caption"><span class="caption-text"><strong>Figure 1:</strong> Simplified annotation workflow.</span><a class="headerlink" href="#id1" title="Permalink to this image">¶</a></p>
</div>
</div>
<div class="section" id="nextflow-pipeline">
<h2>NextFlow pipeline<a class="headerlink" href="#nextflow-pipeline" title="Permalink to this headline">¶</a></h2>
<p>The annotation pipeline summarized in <strong>Figure 1</strong> was fully automated using <a class="reference external" href="https://www.nextflow.io/">NextFlow</a> .
The code with detailed software versions and command line parameters is available on github <a class="reference external" href="https://github.com/metagenlab/annotation_pipeline_nextflow/blob/master/annotation_pipeline.nf">here</a>.</p>
</div>
<div class="section" id="selection-of-pvc-genomes-integrated-into-the-database">
<h2>Selection of PVC genomes integrated into the database<a class="headerlink" href="#selection-of-pvc-genomes-integrated-into-the-database" title="Permalink to this headline">¶</a></h2>
<p>They are more than 2’000 genome assemblies classified as part of the PVC superphylum (NCBI Taxid 1783257)
on the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Undef&amp;id=1783257&amp;lvl=3&amp;p=gcassembly&amp;lin=f&amp;keep=1&amp;srchmode=1&amp;unlock">NCBI taxonomy website</a>. Those genomes are highly redundant (with a high number of genomes of the
same species, in particular for species of the <em>Chlamydia</em> genus). In order to limit the size of the database,
not all genome assemblies were included into the database.
<code class="docutils literal notranslate"><span class="pre">ChlamDB</span> <span class="pre">version</span> <span class="pre">2.0</span></code> (June 2019) includes:</p>
<blockquote>
<div><ul class="simple">
<li><p>All complete PVC genomes (June 2019).</p></li>
<li><p>All draft <em>Chlamydiae</em> genomes excluding draft genomes from species of
the <em>Chlamydia</em> genus with at least one complete genome (June 2019).</p></li>
</ul>
</div></blockquote>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Draft genomes of the most studied <em>Chlamydia</em> species were excluded because they are highly redundant</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Draft genomes of <em>Planctomycetes</em>, <em>Lentisphaerae</em>, <em>Verrucomicrobia</em> and <em>Kiritimatiellaeota</em> phyla were
excluded to limit the total number of genomes in the database. Additional genomes of clades of interest
might be integrated in future releases of the database.</p>
</div>
<p>The list of genomes is available on <a class="reference external" href="https://chlamdb.ch/#genomes">ChlamDB home page</a>. Complete genomes can be distinguished from draft
genomes based on the number of contigs (“N. contigs” column) since draft assemblies exhibit more than one contig.</p>
</div>
<div class="section" id="protein-annotation">
<h2>Protein annotation<a class="headerlink" href="#protein-annotation" title="Permalink to this headline">¶</a></h2>
<p>Protein sequences were annotated using data from multiple databases and various bioinformatic softwares (<em>Figure 1</em>).
Those annotations are summarized on the “locus” page available for each protein (see for instance the entry of <a class="reference external" href="https://chlamdb.ch/locusx?accession=CT_495">CT_495</a>).</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Transporters</span></code> were annotated using the <a class="reference external" href="http://www.tcdb.org/">Transporter Classification Database</a> (TCDB) and the gBLAST3 software from
the <a class="reference external" href="https://github.com/SaierLaboratory/BioVx">BioV suite</a>. Detailed scores including the <em>E-value</em>, <em>score</em>, <em>percentage of identity</em>, the
<em>number of transmembrane domains</em> identified in the query and the reference protein, the query and hit <em>coverage</em>
are reported on the <a class="reference external" href="https://chlamdb.ch/locusx?accession=CT_495">locus page</a>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">COG</span></code> annotation was done with <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK279690/">rpsblast</a> (with an e-value cutoff of <em>0.001</em>) using <a class="reference external" href="https://www.ncbi.nlm.nih.gov/Structure/cdd/cdd_help.shtml#CD_PSSM">position-specific scoring matrix</a>
(PSSM) from the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/Structure/cdd/cdd.shtml">Conserved Domain Database</a> (CDD)</p></li>
<li><dl class="simple">
<dt>Proteins were assigned to <a class="reference external" href="https://www.genome.jp/kegg/ko.html">Kegg Orthologs</a> (KO) numbers with <a class="reference external" href="ftp://ftp.genome.jp/pub/tools/kofamscan/">KofamScan</a> (default parameters) and <a class="reference external" href="https://www.genome.jp/tools/kofamkoala/">KOfam</a> (a database of Hidden Markov Models [HMM] of <a class="reference external" href="https://www.genome.jp/kegg/ko.html">KEGG Orthologs</a>). KofamScan rely on <a class="reference external" href="http://hmmer.org/">HMMER</a> to search the HMM database.</dt><dd><ul>
<li><p>The <a class="reference external" href="https://www.kegg.jp/kegg/rest/keggapi.html">KEGG API</a> was used to retrieve the annotation of <code class="docutils literal notranslate"><span class="pre">Kegg</span> <span class="pre">Orthologs</span></code> and mapping to <code class="docutils literal notranslate"><span class="pre">KEGG</span> <span class="pre">modules</span></code>, <code class="docutils literal notranslate"><span class="pre">KEGG</span> <span class="pre">pathways</span></code> and <a class="reference external" href="https://www.qmul.ac.uk/sbcs/iubmb/enzyme/">Enzyme Commission numbers</a>.</p></li>
</ul>
</dd>
</dl>
</li>
<li><p><a class="reference external" href="https://www.qmul.ac.uk/sbcs/iubmb/enzyme/">Enzyme Commission numbers</a> were assigned based on <a class="reference external" href="https://www.genome.jp/kegg/ko.html">Kegg Orthologs</a> annotations and with <a class="reference external" href="http://priam.prabi.fr/">PRIAM</a>, a method for automated enzyme detection in fully sequenced genomes. PRIAM uses <a class="reference external" href="https://www.ncbi.nlm.nih.gov/Structure/cdd/cdd_help.shtml#CD_PSSM">position-specific scoring matrix</a> (PSSM) of all sequences available in the <a class="reference external" href="https://enzyme.expasy.org/">ENZYME</a> database.</p></li>
<li><dl class="simple">
<dt><a class="reference external" href="https://github.com/ebi-pf-team/interproscan">InterProScan</a> was used to identify protein signatures such as <code class="docutils literal notranslate"><span class="pre">transmembrane</span> <span class="pre">domains</span></code>, <code class="docutils literal notranslate"><span class="pre">Pfam</span> <span class="pre">domains</span></code>, <code class="docutils literal notranslate"><span class="pre">CDD</span> <span class="pre">domains</span></code>,… Detailed InterPro annotations as well as scores associated with each annotation are reported on the “InterproScan” tab of each <a class="reference external" href="https://chlamdb.ch/locusx?accession=CT_495">locus page</a>.</dt><dd><ul>
<li><p><a class="reference external" href="https://github.com/ebi-pf-team/interproscan/wiki/LocalLookupService">The InterProScan match lookup service</a> was used to retrieve pre-calculated InterProScan results for protein integrated into the InterPro database. For sequences not in the lookup service, InterProScan was used to analyse sequences from scratch.</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>Mapping to <a class="reference external" href="https://www.string-db.org/">STRING</a> (a database of protein-protein interactions) was done by exact protein sequence match.
Protein sequences from SRING were indexed using <a class="reference external" href="https://biopython.org/DIST/docs/api/Bio.SeqUtils.CheckSum-module.html">SEGUID checksums</a>. Associations to the scientific literature (<code class="docutils literal notranslate"><span class="pre">PMID</span></code> tab)
were retrived using the <a class="reference external" href="http://version11.string-db.org/help/api/">STRING API</a>.</p></li>
<li><p>Protein subcellular localization predictions were done using <a class="reference external" href="https://www.psort.org/psortb/index.html">PSORTb</a> (withe the <code class="docutils literal notranslate"><span class="pre">--negative</span></code> argument).</p></li>
<li><dl class="simple">
<dt>Effectors of the type III secretion system were predicted using four different softwares:</dt><dd><ul>
<li><p><a class="reference external" href="https://academic.oup.com/bioinformatics/article/27/6/777/235789">BPBAac</a> (default parameters)</p></li>
<li><p><a class="reference external" href="https://academic.oup.com/nar/article/39/suppl_1/D591/2507337">effectiveT3</a> (model <code class="docutils literal notranslate"><span class="pre">TTSS_STD-2.0.2</span></code>, cutoff of <code class="docutils literal notranslate"><span class="pre">0.9999</span></code>)</p></li>
<li><p><a class="reference external" href="https://academic.oup.com/bioinformatics/article/35/12/2051/5165378">DeepT3</a>  (default parameters)</p></li>
<li><p><a class="reference external" href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0058173">T3_MM</a> (default parameters)</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>UNIPARC - SwissProt</p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="homology-search-refseq-and-swissprot">
<h2>Homology search (RefSeq and SwissProt)<a class="headerlink" href="#homology-search-refseq-and-swissprot" title="Permalink to this headline">¶</a></h2>
<p>The closest identifiable RefSeq homologs were</p>
</div>
<div class="section" id="identification-of-orthogroups">
<h2>Identification of Orthogroups<a class="headerlink" href="#identification-of-orthogroups" title="Permalink to this headline">¶</a></h2>
<p>Orthogroups were identified with <a class="reference external" href="https://github.com/davidemms/OrthoFinder">OrthoFinder</a>. This tools identify orthogroups based on BLASTp (evalue cutoff of 0.001)
results using the <a class="reference external" href="https://micans.org/mcl/">MCL</a> software.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><code class="docutils literal notranslate"><span class="pre">Orthologs</span></code> are pairs of genes that descended from a single gene in the last common ancestor (LCA) of two species.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>An <code class="docutils literal notranslate"><span class="pre">orthogroup</span></code> is the group of genes descended from a single gene in the last common ancestor (LCA) of a group of species.
As gene duplication and loss occur frequently in bacteria, we rarely have exactly one ortholog in each considered genome.</p>
</div>
</div>
<div class="section" id="orthogroup-alignments-phylogenies">
<h2>Orthogroup alignments &amp; phylogenies<a class="headerlink" href="#orthogroup-alignments-phylogenies" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="calculation-of-parwise-protein-identity">
<h2>Calculation of parwise protein identity<a class="headerlink" href="#calculation-of-parwise-protein-identity" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="circular-genome-plots-circos">
<h2>Circular genome plots (Circos)<a class="headerlink" href="#circular-genome-plots-circos" title="Permalink to this headline">¶</a></h2>
<div class="align-center figure align-default" id="id2">
<a class="reference internal image-reference" href="../_images/circos_method.png"><img alt="../_images/circos_method.png" src="../_images/circos_method.png" style="width: 400px;" /></a>
<p class="caption"><span class="caption-text">Figure 1: Simplified annotation workflow.</span><a class="headerlink" href="#id2" title="Permalink to this image">¶</a></p>
</div>
</div>
<div class="section" id="species-phylogeny">
<h2>Species phylogeny<a class="headerlink" href="#species-phylogeny" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="comparative-analyses-of-orthogroups-and-cog-kegg-interpro-annotations">
<h2>Comparative analyses of orthogroups and COG/KEGG/Interpro annotations<a class="headerlink" href="#comparative-analyses-of-orthogroups-and-cog-kegg-interpro-annotations" title="Permalink to this headline">¶</a></h2>
<p>Orthogroups KEGG, COG, Interpro entries</p>
</div>
<div class="section" id="prediction-of-protein-protein-interactions">
<h2>Prediction of protein-protein interactions<a class="headerlink" href="#prediction-of-protein-protein-interactions" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="prediction-of-candidate-type-iii-secetion-system-effectors">
<h2>Prediction of candidate type III secetion system effectors<a class="headerlink" href="#prediction-of-candidate-type-iii-secetion-system-effectors" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="taxonomic-profile-of-pfam-domains-and-cogs">
<h2>Taxonomic profile of Pfam domains and COGs<a class="headerlink" href="#taxonomic-profile-of-pfam-domains-and-cogs" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="sql-database-and-web-interface">
<h2>SQL database and Web Interface<a class="headerlink" href="#sql-database-and-web-interface" title="Permalink to this headline">¶</a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Thee SQL database is about 80GB in size. Protein alignments are not stored into the database.</p>
</div>
</div>
<div class="section" id="source-databases">
<h2>Source databases<a class="headerlink" href="#source-databases" title="Permalink to this headline">¶</a></h2>
<table class="docutils align-center" id="id3">
<caption><span class="caption-text">Version of source databases used for the annotation</span><a class="headerlink" href="#id3" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 45%" />
<col style="width: 55%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Database name</p></th>
<th class="head"><p>Version</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>UniprotKB-UNIPARC</p></td>
<td><p>2019.03</p></td>
</tr>
<tr class="row-odd"><td><p>TCDB</p></td>
<td><p>2019.06</p></td>
</tr>
<tr class="row-even"><td><p>RefSeq</p></td>
<td><p>90</p></td>
</tr>
<tr class="row-odd"><td><p>CDD (COG)</p></td>
<td><p>v3.17</p></td>
</tr>
<tr class="row-even"><td><p>InterProScan</p></td>
<td><p>5.35-74.0</p></td>
</tr>
<tr class="row-odd"><td><p>STRING</p></td>
<td><p>v11</p></td>
</tr>
<tr class="row-even"><td><p>PDB</p></td>
<td><p>2019.06</p></td>
</tr>
<tr class="row-odd"><td><p>COG/CDD</p></td>
<td><p>v3.17</p></td>
</tr>
<tr class="row-even"><td><p>KoFam</p></td>
<td><p>2019.04.09</p></td>
</tr>
<tr class="row-odd"><td><p>PRIAM</p></td>
<td><p>2018.06</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="software-versions">
<h2>Software versions<a class="headerlink" href="#software-versions" title="Permalink to this headline">¶</a></h2>
<table class="docutils align-center" id="id4">
<caption><span class="caption-text">Version of the main softwares used for protein annotation</span><a class="headerlink" href="#id4" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 52%" />
<col style="width: 48%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Software name</p></th>
<th class="head"><p>Version</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>FastTree</p></td>
<td><p>2.1.10</p></td>
</tr>
<tr class="row-odd"><td><p>Diamond</p></td>
<td><p>0.9.24</p></td>
</tr>
<tr class="row-even"><td><p>OrthoFinder</p></td>
<td><p>2.2.7</p></td>
</tr>
<tr class="row-odd"><td><p>BLAST</p></td>
<td><p>2.7.1</p></td>
</tr>
<tr class="row-even"><td><p>CheckM</p></td>
<td><p>1.0.12</p></td>
</tr>
<tr class="row-odd"><td><p>KoFamScan</p></td>
<td><p>2019/4/9</p></td>
</tr>
<tr class="row-even"><td><p>Mafft</p></td>
<td><p>7.407</p></td>
</tr>
<tr class="row-odd"><td><p>PSORTb</p></td>
<td><p>3.0.6</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="code-availability">
<h2>Code availability<a class="headerlink" href="#code-availability" title="Permalink to this headline">¶</a></h2>
<table class="docutils align-default">
<colgroup>
<col style="width: 39%" />
<col style="width: 61%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>Website interface</p></td>
<td><p><a class="reference external" href="https://github.com/metagenlab/chlamdb">https://github.com/metagenlab/chlamdb</a></p></td>
</tr>
<tr class="row-even"><td><p>Annotation pipeline</p></td>
<td><p><a class="reference external" href="https://github.com/metagenlab/annotation_pipeline_nextflow">https://github.com/metagenlab/annotation_pipeline_nextflow</a></p></td>
</tr>
<tr class="row-odd"><td><p>Public database download and indexing</p></td>
<td><p><a class="reference external" href="https://github.com/metagenlab/annotation_pipeline_snakemake">https://github.com/metagenlab/annotation_pipeline_snakemake</a></p></td>
</tr>
</tbody>
</table>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Methods</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#nextflow-pipeline">NextFlow pipeline</a></li>
<li><a class="reference internal" href="#selection-of-pvc-genomes-integrated-into-the-database">Selection of PVC genomes integrated into the database</a></li>
<li><a class="reference internal" href="#protein-annotation">Protein annotation</a></li>
<li><a class="reference internal" href="#homology-search-refseq-and-swissprot">Homology search (RefSeq and SwissProt)</a></li>
<li><a class="reference internal" href="#identification-of-orthogroups">Identification of Orthogroups</a></li>
<li><a class="reference internal" href="#orthogroup-alignments-phylogenies">Orthogroup alignments &amp; phylogenies</a></li>
<li><a class="reference internal" href="#calculation-of-parwise-protein-identity">Calculation of parwise protein identity</a></li>
<li><a class="reference internal" href="#circular-genome-plots-circos">Circular genome plots (Circos)</a></li>
<li><a class="reference internal" href="#species-phylogeny">Species phylogeny</a></li>
<li><a class="reference internal" href="#comparative-analyses-of-orthogroups-and-cog-kegg-interpro-annotations">Comparative analyses of orthogroups and COG/KEGG/Interpro annotations</a></li>
<li><a class="reference internal" href="#prediction-of-protein-protein-interactions">Prediction of protein-protein interactions</a></li>
<li><a class="reference internal" href="#prediction-of-candidate-type-iii-secetion-system-effectors">Prediction of candidate type III secetion system effectors</a></li>
<li><a class="reference internal" href="#taxonomic-profile-of-pfam-domains-and-cogs">Taxonomic profile of Pfam domains and COGs</a></li>
<li><a class="reference internal" href="#sql-database-and-web-interface">SQL database and Web Interface</a></li>
<li><a class="reference internal" href="#source-databases">Source databases</a></li>
<li><a class="reference internal" href="#software-versions">Software versions</a></li>
<li><a class="reference internal" href="#code-availability">Code availability</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="../index.html"
                        title="previous chapter">ChlamDB documentation</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="../tutorial/website.html"
                        title="next chapter">Website tutorial</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/methods/annotation.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../tutorial/website.html" title="Website tutorial"
             >next</a> |</li>
        <li class="right" >
          <a href="../index.html" title="ChlamDB documentation"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">ChlamDB 2.0 (June 2019) documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Trestan Pillonel.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 2.2.0.
    </div>
  </body>
</html>