<!DOCTYPE html>


<html>
    <head>
        {% load static %}
        {% load custom_tags %}
        {% load crispy_forms_tags %}
        {% include "chlamdb/header.html" %}
    </head>


    <div class="container-fluid" id="main_container">
      <div class="row">
        <div id="wrapper">
            <div id="page-content-wrapper">
              <div class="row">
                <div class="col-lg-12">
                   {% include "chlamdb/menu.html" %}
                   <div class="home-title"><b> Kegg Orthologs </b></div>
                   <hr class="lines-home">
                    <nav>
                        <ul id="tabs_main" class="nav nav-tabs">
                            <li class="active"><a href="{% url 'extract_ko' %}">Detailed comparison</a></li>
                            <li><a href="{% url 'ko_comparison' %}">Compare KO counts</a></li>
                            <li><a href="{% url 'module_barchart' %}">KO categories barchart</a></li>
                            <li><a href="{% url 'venn_ko' %}">Venn KO</a></li>
                            <li><a href="{% url 'plot_heatmap' 'ko' %}">Whole proteomes heatmaps</a></li>
                            <li><a href="{% url 'pan_genome' 'ko' %}">Pan/Core genome plots</a></li>
                        </ul>
                    </nav>

                     <div id="extract_form">

                        <div class="panel panel-success" style="width:80% ; top: 200px; margin: 10px 10px 10px 10px">
                            <div class="panel-heading" style="width:100%">
                                <h3 class="panel-title">Help</h3>
                            </div>
                            <p style="margin: 10px 10px 10px 10px; line-height: 160%">Get list of Kegg Orthologs shared between genomes selected in panel A and absent in genomes of panel B. Based on <a href="http://www.kegg.jp/ghostkoala/">GhostKOALA</a>.<br>
                            Reference: one of the genome in panel A can be chosen as reference for display the result KO list on a circular map of the genome. Click on "show on circular map" to generate the circos map. Corresponding locus are highlighted in pink. <br>
                            Missing data: if more than one panel A genome (include) is selected, possibility to tolerate missing data in 1 up to 10 genomes (in order to get KO present in at least x out of the y genomes selected in panel A).</p>
                            </p>
                        </div>

                        {% block content %}
						{% csrf_token %}
						{% crispy form %}
                        {% endblock %}
                    </div>



            {% if wrong_n_missing%}

            <div class="panel panel-warning" style="width:500px ; top: 200px; margin: 10px 10px 10px 10px">
                <div class="panel-heading" style="width:100%">
                    <h3 class="panel-title">Help</h3>
                </div>
                <p style="margin: 10px 10px 10px 10px">You cannot set a number of missing data bigger than the number of included genomes</p>
            </div>

            {% elif no_match%}

            <div class="panel panel-warning" style="width:500px ; top: 200px; margin: 10px 10px 10px 10px">
                <div class="panel-heading" style="width:100%">
                    <h3 class="panel-title"></h3>
                </div>
                <p style="margin: 10px 10px 10px 10px">No match</p>
            </div>

            {% endif %}

            {% if envoi_extract %}

                <div id="my-tab-content" class="tab-content">

                    <div class="tab-pane active" id="red">

                        <h3>Number of Kegg Orthologs: {{ sum_group }} </h3>

                        <table class="display" id="extract_ko">

                            <thead>
                                <tr>
                                    <th id="entete_locus">KO</th>
                                    <th>Description</th>
                                    <th>Kegg Pathways</th>
                                    <th>Kegg Modules</th>
                                    <th>present in (/{{include|length}})</th>
                                    <th>freq complete database {{max_n}}</th>
                                </tr>
                            </thead>

                            {% for value in match_groups_data %}
                                <tr>
                                    <td><a href="{% url 'fam_ko' value.0 %}">{{value.0}}</a></td>
                                    <td>{{value.1}}</td>
                                    <td>{{value.2|safe}}</td>
                                    <td>{{value.3|safe}}</td>
                                    <td>{{value.4}}</td>
                                    <td>{{value.5}}</td>
                                </tr>
                            {% endfor %}

                        </table>
                    </div>
				  </div>
			  {% endif %}
		  </div>
          </div>
        </div>
      </div>
    </div>
  </div>

<script type="text/javascript" id="js2">
$(document).ready(function() {
	$('#extract_ko').DataTable( {
		dom: 'Bfrtip',
		"paging":   true,
		"ordering": true,
		"info":     false,
		buttons: [{extend: 'excel', title: 'ko_table'},
		{ extend: 'csv', title: 'ko_table'}],
	});
});
</script>

</body>
{% include "chlamdb/style_menu.html" %}
{% include "chlamdb/show_hide_plasmid_accessions.html" %}

<style>
#div_id_checkbox_single_copy{
  display: none;
}
</style>

</html>
