<!DOCTYPE html>


<html>
    <head>
        {% load custom_tags %}
        {% load static %}
        {% load static %}
        {% load crispy_forms_tags %}
        {% include "chlamdb/header.html" %}
    </head>

    <div class="container-fluid" id="main_container">
      <div class="row">
        <div id="wrapper">
            <div id="page-content-wrapper">
              <div class="row">
                <div class="col-lg-12">
                   {% include "chlamdb/menu.html" %}

                    <nav>
                        <ul id="tabs_main" class="nav nav-tabs">
                            <li class="active"><a href="{% url 'extract_interpro' %}">Detailed comparison</a></li>
                            <li><a href="{% url 'venn_interpro' %}">Venn diagram</a></li>
                            <li><a href="{% url 'plot_heatmap' 'interpro' %}">Whole proteomes heatmaps</a></li>
                            <li><a href="{% url 'pan_genome' 'interpro' %}">Pan/Core genome plots</a></li>
                            <li><a href="{% url 'interpro_taxonomy' %}">Interpro taxonomy</a></li>
                        </ul>
                    </nav>

                     <div id="extract_form">
                        <div class="panel panel-success" style="width:80% ; top: 200px; margin: 10px 10px 10px 10px">
                            <div class="panel-heading" style="width:100%">
                                <h3 class="panel-title">Help</h3>
                            </div>
                            <p style="margin: 10px 10px 10px 10px; line-height: 160%">Get list of Interpro entries shared between genomes selected in panel A and absent in genomes of panel B. Based on InterProScan-5.18-57.0. <br>
                            Reference: one of the genome in panel A can be chosen as reference for display the result Interpro entries list on a circular map of the genome. Click on "show on circular map" to generate the circos map. Corresponding locus are highlighted in pink. <br>
                            Missing data: if more than one panel A genome (include) is selected, possibility to tolerate missing data in 1 up to 10 genomes (in order to get Interpro entries present in at least x out of the y genomes selected in panel A).</p>
                        </div>


                        {% block content %}
                        <form action="{% url "extract_interpro" %}" method="post" id="extract-form">
                            {% csrf_token %}
                            {% crispy form %}
                        </form>
                        {% endblock %}
                    </div>

                    <div class="col-lg-12">
                            <div id="progress-div" class="row" style="padding-top:5px;height:25px;" class="row"></div>
          
                            <div class="progress row" style="width=50%;" id="progress">
                              <div class="progress-bar progress-bar-striped active" role="progressbar" aria-valuenow="0" aria-valuemin="0" aria-valuemax="100" style="width: 20%"></div>
                            </div>
                            <div id="results" class="row"></div>
                      </div>

                </div>
            </div>
          </div>
        </div>
      </div>
    </div>



    <div id="footer">

    </div>

<script type="text/javascript">
    $("#progress").hide();

    var frm = $('#extract-form');
    var prg_dv = $('#progress-div');
    var pgrbar = $('.progress-bar');
    frm.submit(function () {
        console.log("submit!!!")
        $.ajax({
            type: frm.attr('method'),
            url: frm.attr('action'),
            data: frm.serialize(),
            success: function (data) {
                console.log("success");
                console.log(data.task_id);
                if (data.task_id != null) {
                    get_task_info(data.task_id);
                }
            },
            error: function (data) {
                console.log("Something went wrong!");
            }
        });
        return false;
    });

    function get_task_info(task_id) {
        $.ajax({
            type: 'get',
            url: '/get-task-info/',
            data: {'task_id': task_id},
            success: function (data) {
                pgrbar.css('width', 10+'%').attr('aria-valuenow', 10)
                $("#progress").show();
                $('#results').html('')
                if (data.state == 'PENDING') {
                    prg_dv.html('Please wait...');
                }
                else if (data.state == 'PROGRESS') {
                    pgrbar.css('width', data.result.percent+'%').attr('aria-valuenow', data.result.percent)
                    console.log(data.result.percent);
                    prg_dv.html('Step ' + data.result.current + '/' + data.result.total + ': ' + data.result.description);
                }
                else if (data.state == 'SUCCESS') {
                    // pgrbar.hide();
                    prg_dv.html('Done');
                    $('#results').html(data.result);
                    $("#progress").hide();

                    

                    $('#circos_form').append('<input type="hidden" name="csrfmiddlewaretoken" value="{{ csrf_token }}">');

                    $(document).ready(function() {

                var x=document.getElementById("id_reference");
                for (var i = 0; i < x.options.length; i++) {
                    $("#id_reference").children("option[value^=" + x.options[i].value + "]").hide()
                }

                $("#id_orthologs_in").change(function() {
                    document.getElementById("id_reference").selectedIndex=0;
                    var x=document.getElementById("id_orthologs_in");
                    for (var i = 0; i < x.options.length; i++) {
                        if(x.options[i].selected ==true){

                        $("#id_reference").children("option[value^=" + x.options[i].value + "]").show()
                        }else{
                        $("#id_reference").children("option[value^=" + x.options[i].value + "]").hide()

                        }
                        }
                })

                })



                $(document).ready(function() {
                $('#interpro_table').DataTable( {
                dom: 'Bfrtip',
                "paging":   true,
                "ordering": true,
                "info":     false,
                buttons: [
                {
                extend: 'excel',
                title: 'cog_table'
                },
                {
                extend: 'csv',
                title: 'cog_table'
                }
                ],
                } );
                } );

                $(document).ready(function() {
                $('#reference_genome').DataTable( {
                dom: 'Bfrtip',
                "paging":   true,
                "ordering": true,
                "info":     false,
                buttons: [
                {
                extend: 'excel',
                title: 'reference_genome_locus_list'
                },
                {
                extend: 'csv',
                title: 'reference_genome_locus_list'
                }
                ],
                } );
                } );


                }

                if (data.state != 'SUCCESS') {
                    setTimeout(function () {
                        get_task_info(task_id)
                    }, 1000);
                }
            },
            error: function (data) {
                prg_dv.html("Something went wrong!");
            }
        });
    }
</script>




    </body>
    <style>
    #div_id_checkbox_single_copy{
      display: none;
    }
    </style>
    {% include "chlamdb/style_menu.html" %}
</html>
