<!DOCTYPE html> <html>  <head> {% load custom_tags %}
    {% load static %}
    {% load static %}
    {% load crispy_forms_tags %}
    {% include "chlamdb/header.html" %}
</head>

<div class="container-fluid" id="main_container">
    <div class="row">
      <div id="wrapper">
        <div id="page-content-wrapper">
          <div class="row">
            <div class="col-lg-12">
             {% include "chlamdb/menu.html" %}
             <div class="home-title"><b> Pfam domains </b></div>
             <hr class="lines-home">

             <nav>
                <ul id="tabs_main" class="nav nav-tabs">
                    <li class="active"><a href="{% url 'extract_pfam' %}">Detailed comparison</a></li>
                    <li><a href="{% url 'venn_pfam' %}">Venn diagram</a></li>
                    <li><a href="{% url 'pfam_comparison' %}">Compare Pfam Domains</a></li>
                    <li><a href="{% url 'plot_heatmap' 'Pfam' %}">Whole proteomes heatmaps</a></li>
                    <li><a href="{% url 'pan_genome' 'Pfam' %}">Pan/Core genome plots</a></li>

                </ul>
            </nav>
            <div class="panel panel-info" style="width:90% ; top: 200px; margin: 10px 10px 10px 10px">
                <div class="panel-heading" style="width:100%; ">
                        <h3 class="panel-title">Info about the analysis    <a href="/docs/index.html" id="show-option" title="See an axample of the analysis you can perform"><i class="fab fa-info-circle fa-2x" style="size: 5em;" ></i></a></h3>

                    </div>
                    <p style="margin: 10px 10px 10px 10px; line-height: 180%">Get the list of Pfam domains shared between genomes (selected in panel A) and absent in others (selected in panel B). <br>
                        It is possible to distinguish between plasmids and chromosomes by clicking on the dedicated checkbox.
                        <br>Missing data: if more than one genome is selected in panel A, it is possible to tolerate missing data in 1 up to 10 genomes. In this way Pfam domains present in at least x out of y genomes selected are kept.</p></p>
                </div>
   

                        {% block content %}
						{% csrf_token %}
						{% crispy form %}
                        {% endblock %}

                    </div>

                    {% if wrong_n_missing %}

                    <div class="panel panel-warning" style="width:500px ; top: 200px; margin: 10px 10px 10px 10px">
                        <div class="panel-heading" style="width:100%">
                            <h3 class="panel-title">Help</h3>
                        </div>
                        <p style="margin: 10px 10px 10px 10px">You cannot set a number of missing data bigger than the number of included genomes</p>
                    </div>

                    {% elif no_match %}

                    <div class="panel panel-warning" style="width:500px ; top: 200px; margin: 10px 10px 10px 10px">
                        <div class="panel-heading" style="width:100%">
                            <h3 class="panel-title"></h3>
                        </div>
                        <p style="margin: 10px 10px 10px 10px">No match</p>
                    </div>

                    {% endif %}


                    {% if envoi_extract %}

                    <div class="row"  style="background-color: rgba(245, 245, 245, 0.986)">
                        <div class="col-lg-12">
                          <ul id="tabs" class="nav nav-tabs" data-tabs="tabs">
                            <br>
                            <center><h3 style="padding:0; color: rgb(0, 0, 0);">Pfam domains shared by the selected genomes </h3></center>
                            <br>
                            <div class="panel panel-success" style="width:60%; top: 200px; margin: 10px 10px 10px 10px; float: left;">
                              <div class="panel-heading" style="width:100%">
                                <h3 class="panel-title">Help to interpret the results</h3>
                              </div>
                              <p style="margin: 10px 10px 10px 10px; line-height: 180%">Two tables have been generated:
                                <br> <b>Pfam</b> table: it contains the list of Pfam domains shared among the selected genomes.Pfam entry, its description, its frequency in the selected genomes and in all genomes is reported. 
                              <br><b>Locus list reference</b>: a complete list of the members of each Pfam group shared by the selected genome is displayed.                            
                              </p>
                            </div>

                            <ul style="list-style-type:square">
                                <div style="padding-top:10px">
   
                                   <table class="table" style="width:600px;">
                                       <tr>
                                           <th>Number of included genomes</th>
                                           <td> {{ include|length }} </td>
                                       </tr>
   
                                       <tr>
                                           <th>Number of excluded genomes</th>
                                           <td> {{ exclude|length }} </td>
                                       </tr>
   
                                        <tr>
                                           <th>Number of accepted missing data</th>
                                           <td> {{ n_missing }} </td>
                                       </tr>
                                         <tr>
                                           <th>Number of  PFAM domains identified</th>
                                           <td> {{ sum_group }}  </td>
                                       </tr>
    
                                   </table>

                        <div class="col-lg-12">
                            <ul id="tabs" class="nav nav-tabs" data-tabs="tabs">
                              <br>
                              <li class="active"><a href="#red" data-toggle="tab">Pfam</a></li>
                              <li><a href="#tab2" data-toggle="tab">Locus list reference</a></li>
                        </ul>
                

                        <div id="my-tab-content" class="tab-content" style="margin-top: 2em;">

                            <div class="tab-pane active" id="red">
                                <table class="display" id="PFAM_table">
                                    <thead>
                                        <tr>
                                        <th></th>
                                        <th id="entete_locus">Pfam entry</th>
                                        <th>Description</th>
                                            <th>Count (/{{n_genomes}})</th>
                                            <th>Count all (/{{max_n}})</th>
                                        </tr>
                                    </thead>

                                    {% for value in match_groups_data %}
                                        <tr>
                                            <td>{{value.0}}</td>
                                            <td><a href="{% url 'fam_pfam' value.1 %}">{{value.1}}</a></td>
                                            <td>{{value.2}}</td>
                                            <td>{{value.3}}</td>
                                            <td>{{value.4}}</td>

                                        </tr>
                                    {% endfor %}

                                </table>
                            </div>
                        </div>

                    {% endif %}
                </div>
          </div>
        </div>
      </div>
    </div>
  </div>

<script type="text/javascript" id="js2">
  $(document).ready(function() {

      var x=document.getElementById("id_reference");
      for (var i = 0; i < x.options.length; i++) {
          $("#id_reference").children("option[value^=" + x.options[i].value + "]").hide()
      }

      $("#id_orthologs_in").change(function() {
          document.getElementById("id_reference").selectedIndex=0;
          var x=document.getElementById("id_orthologs_in");
          for (var i = 0; i < x.options.length; i++) {
              if(x.options[i].selected ==true){

              $("#id_reference").children("option[value^=" + x.options[i].value + "]").show()
              }else{
              $("#id_reference").children("option[value^=" + x.options[i].value + "]").hide()

              }
              }
      })

  })


$(document).ready(function() {
$('#PFAM_table').DataTable( {
dom: 'Bfrtip',    
"paging":   true,
"ordering": true,
"info":     false,
buttons: [
{
    extend: 'excel',
    title: 'PFAM_table'
},
{
    extend: 'csv',
    title: 'PFAM_table'
}
],
} );
} );

$(document).ready(function() {
$('#reference_genome').DataTable( {
 dom: 'Bfrtip',   
"paging":   true,
"ordering": true,
"info":     false,
buttons: [
{
    extend: 'excel',
    title: 'reference_genome'
},
{
    extend: 'csv',
    title: 'reference_genome'
}
],
} );
} );


 type="text/javascript" id="js">
            $(document).ready(function() {
            // call the tablesorter plugin
            $("table").tablesorter({
            // change the default sorting order from 'asc' to 'desc'
            sortInitialOrder: 1
            });
            });
        </script>
        <script type="text/javascript">
            jQuery(document).ready(function ($) {
                $('#tabs_main').tab();
            });
        </script>
        <script type="text/javascript">
            jQuery(document).ready(function ($) {
                $('#tabs1').tab();
            });
       
$(document).ready(function() {
$("#PFAM_table").DataTable();} );
$(document).ready(function() {
$("#tab2").DataTable();} );

</script>
</body>
{% include "chlamdb/style_menu.html" %}
{% include "chlamdb/show_hide_plasmid_accessions.html" %}
<style>

#div_id_checkbox_single_copy{
  display: none;
}
</html>
