<!DOCTYPE html>


<html>

<head>
{% load static %}
{% load static %}
{% load custom_tags %}
{% get_static_prefix as STATIC_PREFIX %}
{% include "chlamdb/header.html" %}
</head>

<body>


{% if not menu %}
<div class="container-fluid" id="container_no_margin" height="100%">

{% else %}
<div class="container-fluid" id="main_container">
{% endif %}

<div class="row">
<div id="wrapper">
  <div id="page-content-wrapper">
	<div class="row">
	  <div class="col-lg-12 col-md-12 col-sm-12">

	  {% if menu %}
		  {% include "chlamdb/menu.html" %}
	  {% endif %}

	  {% if not valid_id %}
		  <div class="alert alert-warning" role="alert" style="width:50%"><a>Unknown accession! Accept protein ids, locus tags and orthogroup IDs</a></div>
		  <p>
	  {% else %}

		<ul class="nav nav-tabs">
			<li class="active">
				<a data-toggle="tab" href="#tab33">Overview</a>
			</li>
			<li>
				<a href="#12" data-toggle="tab">Homologs <span class="badge">{{n_homologues}}</span></a>
			</li>
			<li>
				<a href="#1" data-toggle="tab">Phylogenetic distribution</a>
			</li>

		{% if homologues > 2 and optional2status|keyvalue:"gene_phylogenies" %}
			<li>
				<a href="#9" data-toggle="tab" onclick="window.open('{% url 'phylogeny' og %}', 'frame9'); return false">Orthogroup phylogeny</a>
			</li>
		{% endif %}
		</ul>

		  <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/plotly.js/1.48.1/plotly.min.js"></script>

	
		  <div class="tab-content">

		  <div class="tab-pane" id="12">
			  {% include "chlamdb/homologs.html" %}
		  </div>

		  <div class="tab-pane active" id="tab33">
			<h3>Orthogroup: {{og}} -- {{n_homologues}} members: overview of the annotation</h3>


			  <div class="panel panel-success" style="width:80%; top: 200px; margin: 10px 10px 10px 10px">
				  <div class="panel-heading" style="width:100%">
					  <h3 class="panel-title">Help</h3>
				  </div>
				  <p style="margin: 10px 10px 10px 10px">Orthologs were identified with <a href="https://github.com/davidemms/OrthoFinder">OrthoFinder</a>. The annotation of orthologous proteins can differ depending the approach used 
				  to annotate the source genomes. Automated annotations are also subject to errors. Spurious clustering of non-orthologous proteins can also happen. 
				  This page summarizes the annotation of all members of the orthologous group <strong>{{locus}}</strong>, allowing to check for eventual inconsistencies. <br>
				  The number of times (occurence) an annotation is observed in the orthogroup is reported for each annotation.<br>
				  </p>
			  </div>

			<div class="col-md-12 col-lg-4">
			  <div class="container-fluid">

				<div class="row">
				  <div class="panel panel-default">
					<div class="panel-heading">
					  <h3 class="panel-title">Gene</h3>
					</div>
						<table class="table table-striped" style="width: 100%">
							<thead>
							  <tr>
								  <th>#</th>
								  <th>Gene name</th>
								  <th>Occurence</th>
							  </tr>
							</thead>

							<tbody>
								{% for row in gene_annotations %}
								  <tr>
									<td>{{row.0}}</td>
									<td>{{row.1}}</td>
									<td>{{row.2}}</td>  
								  </tr>
								{% endfor %}
							</tbody>
						</table>
				  </div>
				</div>



					<div class="row">
					  <div class="panel panel-default">
						<div class="panel-heading">
						  <h3 class="panel-title">UniProtKB annotation</h3>
						</div>
							<table class="table table-striped" style="width: 100%">
								<thead>
								  <tr>
									  <th></th>
									  <th>Mapped</th>
									  <th>Unmapped</th>
								  </tr>
								</thead>
								<tbody>
									  <tr>
										<td></td>
										<td>{{mapped}}</td>
										<td>{{unmapped}}</td>  
									  </tr>
								</tbody>
							</table>


						<div class="panel-heading">
						  <h3 class="panel-title">{{reviewed|length}} reviewed SwissProt entries</h3>
						</div>
						{% if reviewed|length != 0 %}
							<table class="table table-striped" style="width: 100%">
								<thead>
									<tr>
										<th>UniProt accession</th>
										<th>Annotation score</th>
										<th>Gene</th>
										<th>Full Name</th>
									</tr>
								</thead>
								<tbody>
								{% for swissprot in reviewed %}
									  <tr>
										  <td><a href="https://www.uniprot.org/uniprot/{{swissprot.0}}">{{swissprot.0}}</a></td>
										  <td>{{swissprot.2}}</td>
										  <td>{{swissprot.3}}</td>  
										  <td>{{swissprot.4}}</td>   
									  </tr>
								{% endfor %}
								</tbody>
							</table>
						{% endif %}
					  </div>
					</div>


					<div class="row">
					  <div class="panel panel-default">
						<div class="panel-heading">
						  <h3 class="panel-title">Product</h3>
						</div>
							<table class="table table-striped" style="width: 100%">
								<thead>
								  <tr>
									  <th>#</th>
									  <th>Product</th>
									  <th>Occurence</th>
								  </tr>
								</thead>
								<tbody>
									{% for row in product_annotations %}
									  <tr>
										<td>{{row.0}}</td>
										<td>{{row.1}}</td>
										<td>{{row.2}}</td>  
									  </tr>
									{% endfor %}
								</tbody>
							</table>
					  </div>
					</div>


				</div>
			  </div>

			  <div class="col-md-12 col-lg-8">
				  <div class="container-fluid">

					 {% if COG_annotations|length > 0 %}
						<div class="row">
						  <div class="panel panel-default">
							<div class="panel-heading">
							  <h3 class="panel-title">COG annotation(s)</h3>
							</div>
								<table class="table table-striped" style="width: 100%">
									<thead>
									  <tr>
										  <th>#</th>
										  <th>COG</th>
										  <th>Occurence</th>
										  <th>Description</th>
										  <th>Category</th>
										  <th>Category description</th>
									  </tr>
									</thead>
									<tbody>
										{% for row in COG_annotations %}
										  <tr>
											<td>{{row.0}}</td>
											<td><a href="{% url 'fam' row.1 'cog' %}">{{row.1}}</a></td>
											<td>{{row.3}}</td> 
											<td>{{row.2}}</td> 
											<td>{{row.4}}</td>
											<td>{{row.5}}</td>
										  </tr>
										{% endfor %}
									</tbody>
								</table>
						  </div>
						</div>
					  {% endif %}


					 {% if KO_annotations|length > 0 %}
						<div class="row">
						  <div class="panel panel-default">
							<div class="panel-heading">
							  <h3 class="panel-title">Kegg Ortholog annotation(s)</h3>
							</div>
							<table class="table table-striped" style="width: 100%">
								<thead>
								  <tr>
									  <th>#</th>
									  <th>KO</th>
									  <th>Occurence</th>
									  <th>Name</th>
									  <th>Definition</th>
									  <th>EC</th>
									  <th>pathways</th>
									  <th>modules</th>
								  </tr>
								</thead>
								<tbody>
									{% for row in KO_annotations %}
									  <tr>
										<td>{{row.0}}</td>
										<td><a href="{% url 'fam' row.1 'ko' %}">{{row.1}}</a></td>
										<td>{{row.2}}</td> 
										<td>{{row.3}}</td> 
										<td>{{row.4}}</td>
										<td>{{row.5}}</td>
										
										<td>
											{% for pathway in row.6 %}
												{% if pathway != '-' %}
													<a href="{% url 'KEGG_mapp_ko' pathway %}">{{pathway}}</a>
												{% else %}
													-
												{% endif %}
											{% endfor %}
										</td>
										<td>
											{% for module in row.7%}
												{% if module != '-' %}
													 <a href="{% url 'KEGG_module_map' module %}">{{module}}</a>
												{% else %}
													-
												{% endif %}
											{% endfor %}
										</td>
									  </tr>
									{% endfor %}
								</tbody>
							</table>
						  </div>
						</div>
					  {% endif %}


					  {% if length_distrib %}
						  <div class="row">
							 <div class="panel panel-default">
							  <div class="panel-heading">
							  	<h3 class="panel-title">Protein length</h3>
							  </div>
							  <table class="table" id="length" style="width:300px">
								  <tbody>

									{% if single_length %}
										<tr>
											<th>Unique protein length </th>
											<td>{{prot_length}}</td>
										</tr>

									{% else %}
										<tr>
											<th>Min length</th>
											<td>{{min_protein_length}}</td>
										</tr>
										 <tr>
											<th>Max length</th>
											<td>{{max_protein_length}}</td>
										</tr>
										<tr>
											<th>Mean length (sd)</th>
											<td>{{mean_protein_length}} ({{std_protein_length}})</td>
										</tr>
										<tr>
											<th>Median length</th>
											<td>{{median_protein_length}}</td>
										</tr>
									{% endif %}

								  </tbody>
							  </table>

							  <div class="panel-heading">
								  <h3 class="panel-title">Protein length distribution</h3>
							  </div>

							  <div style="width:80%">
							       {{html_plot_prot_length|safe}}
							  </div>
						   </div>
						</div>
                  	{% endif %}
				  </div>
				  </div>
              {% endif %}
          </div>
		  </div>
		  </div>
        </div>
      </div>
    </div>
  </div>
</div>
{% include "chlamdb/style_menu.html" %}
<style>
#description_field{
    width:100px;
}

</style>
</body>
</html>
