<!DOCTYPE html>


<html>



<head>
    {% load staticfiles %}
    {% load static %}
    {% get_static_prefix as STATIC_PREFIX %}
    {% load custom_tags %}
    {% include "chlamdb/header.html" %}
</head>


<body>

  <div class="container-fluid" id="main_container">
    <div class="row">
      <div id="wrapper">
          <div id="page-content-wrapper">
            <div class="row">
              <div class="col-lg-12">
                 {% include "chlamdb/menu.html" %}
                  <nav>
                      <ul id="tabs_main" class="nav nav-tabs">
                          <li><a href="{% url 'extract_orthogroup'  %}">Detailed comparison</a></li>
                          <li><a href="{% url 'venn_orthogroup'  %}">Venn diagram</a></li>
                          <li class="active"><a href="{% url 'orthogroup_comparison'  %}">Compare Orthogroups size</a></li>
                          <li><a href="{% url 'plot_heatmap'  'orthology' %}">Whole proteomes heatmaps</a></li>
                          <li><a href="{% url 'pan_genome'  'orthology' %}">Pan/Core genome plots</a></li>

                      </ul>
                  </nav>

                      <div class="panel panel-success" style="width:80% ; top: 200px; margin: 10px 10px 10px 10px">
                          <div class="panel-heading" style="width:100%">
                              <h3 class="panel-title">Help</h3>
                          </div>
                          <p style="margin: 10px 10px 10px 10px; line-height: 160%">Compare orthogroup presenting more than one homolog in one of the selected genomes.</p>
                      </p>
                      </div>



                      <form action='{% url "orthogroup_comparison"  %}' method="post">
                          {% csrf_token %}
                          {{ form.as_p }}
                          <input type="submit" value="Submit" />

                      </form>

                      {% if envoi_comp %}

                      <div class="panel panel-success" style="width:80% ; top: 200px; margin: 10px 10px 10px 10px">
                          <div class="panel-heading" style="width:100%">
                              <h3 class="panel-title">Help</h3>
                          </div>
                          <p style="margin: 10px 10px 10px 10px; line-height: 160%">Homolog counts can be reordrered by clicking on column headers.<br>
                          <strong>Total count:</strong> number of homologs in the whole database </p>
                      </p>
                      </div>

                              Number of orthogroups exhiniting > 1 copy (in one ore multiple genomes) in at least 1 out of {{taxon_list|length}} selected genomes: {{orthogroups2counts|length}}

                              <table class="sortable" id="mytable">
                                  <style>
                                  table {
                                  width:100%;
                                  table-layout:fixed;
                                  }
                                  td {
                                  overflow: hidden;
                                  white-space: nowrap;
                                  }
                                  </style>
                                  <thead>


                                    <tr style="height:200px;">
                                        <th valign="bottom">Orthogroup</th>
                                        <th valign="bottom">Total count</th>
                                        {% for taxon in taxon_list %}
                                            <th id="entete_meta4" valign="bottom"><div class="rotate">{{ taxon_id2description|keyvalue:taxon}}</div></th>
                                        {% endfor %}
                                        <th valign="bottom">Annot</th>
                                    </tr>
                                  </thead>
                                  <tbody>
                                  {% for key, values in orthogroups2counts.items %}
                                      <tr> <td><a href="{% url 'locusx'  key True%}">{{key}}</a></td>
                                          <td>{{orthogroups2total_count|keyvalue:key}}</td>
                                          {% for taxon_count in values%}
                                                {%if taxon_count != 0 %}
                                                <td data-hist="{{taxon_count}}">{{taxon_count}}</td>
                                                {% else %}
                                                 <td>{{taxon_count}}</td>
                                                {% endif %}
                                          {% endfor %}
                                          <td>
                                          {% for values in group2annot|keyvalue:key %}
                                              {{values.0}}({{values.1}})<br/>
                                          {%endfor%}
                                          </td>
                                      </tr>
                              {% endfor %}
                              </tbody>
                              </table>



                      {% endif %}
              </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>

<script src="{% static 'js/jquery.hottie.js' %}"></script>
<script>
$(function(){
$("#mytable td").hottie({
readValue : function(e) {
return $(e).attr("data-hist");
},
colorArray : [
"#5858FA",
"#FBEFEF",
"#FA5858"
]

});

});
</script>


<script>

$(document).ready(function() {
  $('#mytable').DataTable( {
    dom: 'lBfrtip',
    buttons: [
        {
            extend: 'colvis',
            columns: ':not(.noVis)',
        },
        {
            extend: 'excel',
            title: 'plast_refseq'
        },
        {
            extend: 'csv',
            title: 'plast_refseq'
        }
    ],
    "paging":   false,
    "ordering": true,
    "info":     false,
    } );
} );

</script>

{% include "chlamdb/style_menu.html" %}
</html>
