<!DOCTYPE html>


<html>
<head>
  {% load static %}
  {% include "chlamdb/header.html" %}
</head>
<script src="https://d3js.org/d3.v4.js"></script>
<div class="container-fluid" id="main_container">
  <div class="row">
    <div id="wrapper">
        <div id="page-content-wrapper">
          <div class="row">
            <div class="col-lg-12">
               {% include "chlamdb/menu.html" %}
               <div class="home-title" ><b>   {% if type == 'Pfam' %} Pfam domains  {% endif %}  {% if type == 'COG' %} COGs  {% endif %} {% if type == 'orthology' %} Orthogroups  {% endif %} {% if type == 'ko' %} Kegg Orthologs {% endif %} </b></div>
               <hr class="lines-home">
                <nav>
                    <ul id="tabs_main" class="nav nav-tabs">

                          {% if type == 'Pfam' %}
                              <li><a href="{% url 'extract_pfam' %}">Detailed comparison</a></li>
                              <li><a href="{% url 'venn_pfam' %}">Venn diagram</a></li>
                              <li><a href="{% url 'pfam_comparison' %}">Compare Pfam Domains</a></li>
                              <li><a href="{% url 'plot_heatmap' 'Pfam' %}">Whole proteomes heatmaps</a></li>
							  <li class="active"><a href="{% url 'pan_genome' 'Pfam' %}">Pan/Core genome plots</a></li>
                          {% elif type == 'COG' %}
                              <li><a href="{% url 'extract_cog' %}">Detailed comparison</a></li>
                              <li><a href="{% url 'venn_cog' %}">Venn diagram</a></li>
                              <li><a href="{% url 'cog_barchart' %}">COG categories barchart</a></li>
                              <li><a href="{% url 'COG_phylo_heatmap' True %}">COG heatmap freq.</a></li>
                              <li><a href="{% url 'COG_phylo_heatmap' False %}">COG heatmap counts</a></li>
                              <li><a href="{% url 'plot_heatmap' 'COG' %}">Whole proteomes heatmaps</a></li>
                              <li class="active"><a href="{% url 'pan_genome' 'COG' %}">Pan/Core genome plots</a></li>
                          {% elif type == 'ko' %}
                              <li><a href="{% url 'extract_ko' %}">Detailed comparison</a></li>
                              <li><a href="{% url 'ko_comparison' %}">Compare KO counts</a></li>
                              <li><a href="{% url 'module_barchart' %}">KO categories barchart</a></li>
                              <li><a href="{% url 'venn_ko' %}">Venn KO</a></li>
                              <li><a href="{% url 'plot_heatmap' 'ko' %}">Whole proteomes heatmaps</a></li>
                              <li class="active"><a href="{% url 'pan_genome' 'ko' %}">Pan/Core genome plots</a></li>
                          {% elif type == 'orthology' %}
                              <li><a href="{% url 'extract_orthogroup' %}">Detailed comparison</a></li>
                              <li><a href="{% url 'venn_orthogroup' %}">Venn diagram</a></li>
                              <li><a href="{% url 'orthogroup_comparison' %}">Compare Orthogroups size</a></li>
                              <li><a href="{% url 'plot_heatmap' 'orthology' %}">Whole proteomes heatmaps</a></li>
                              <li class="active"><a href="{% url 'pan_genome' 'orthology' %}">Pan/Core genome plots</a></li>
                          {% endif %}
                  </ul>
                  </nav>


                  <div class="panel panel-success" style="width:80%; top: 200px; margin: 10px 10px 10px 10px">
                      <div class="panel-heading" style="width:100%">
                          <h3 class="panel-title">Help</h3>
                      </div>
                      <p style="margin: 10px 10px 10px 10px">Accumulation/rarefaction plot of

                          {% if type == 'Pfam' %}
                              Pfam domains
                          {% elif type == 'COG' %}
                              COG
                          {% elif type == 'ko' %}
                              Kegg orthologs
                          {% elif type == 'orthology' %}
                              Orthologous groups (pan/core genome plots)
                          {% endif %}
                          <br> Targets = 'Select all' to get the core
                          {% if type == 'Pfam' %}
                              Pfam domains
                          {% elif type == 'COG' %}
                              COG
                          {% elif type == 'ko' %}
                              Kegg orthologs
                          {% elif type == 'orthology' %}
                              Orthologous groups
                          {% endif %}
                           of the entire dataset

                          <br>
                      </p>
                  </div>

              <form action='{% url "pan_genome" type %}' method="post">
                  {% csrf_token %}
                  {{ form.as_p }}
                  <input type="submit" value="Submit" />

              </form>
              </div>

            {% if envoi %}
			  <div class="tab-pane active" id="tab1">
				<div class="panel panel-success" style="width:80%; top: 200px; margin: 10px 10px 10px 10px">
					<div id="rarefaction_plot">
					</div>
				</div>
			  </div> <!-- tab -->

           {% endif %}
         </div>
       </div>
      </div>
    </div>
  </div>
</div>


{% include "chlamdb/style_menu.html" %}
</body>

<script>

var image_size_x = 900;

var legend_spacing_x = 20;
var legend_size_x = 250;

var data_count = {{ js_data_count |safe }};
var data_acc = {{ js_data_acc|safe }};
var margins = {top: 10, bottom:80, left:100, right:10};
var width = image_size_x-margins.left-margins.right-legend_size_x;
var height = 500-margins.bottom-margins.top;

var min_val = d3.min(data_count);
var max_val = d3.max(data_acc);

var acc_color="#69b3a2";
var shared_color="#FF6347";

var x_scale = d3.scaleLinear().
	domain([1, data_count.length]).
	range([0, width]);

var y_scale = d3.scaleLinear().
	domain([0.95*min_val, 1.05*max_val]).
	range([height, 0]);

var svg = d3.select("#rarefaction_plot").
	append("svg").
	attr("width", image_size_x).
	attr("height", height+margins.bottom+margins.top).
	append("g").
	attr("transform",
		"translate(" + margins.left + "," + margins.top +")");

// var legend_position_x = ;
var legend_position_y = 15;

// construct the legend
svg.append("circle").
	attr("cx", width+legend_spacing_x).
	attr("cy", height/2-legend_position_y).
	attr("r", 4).
	attr("fill", acc_color);

svg.append("text").
	attr("x", width+legend_spacing_x+10).
	attr("y", height/2-legend_position_y+10-6).
	text("pan-genome size");

svg.append("circle").
	attr("cx", width+legend_spacing_x).
	attr("cy", height/2+legend_position_y).
	attr("r", 4).
	attr("fill", shared_color);

svg.append("text").
	attr("x", width+legend_spacing_x+10).
	attr("y", height/2+legend_position_y+10-6).
	text("{{type_txt}} shared by n genomes");

svg.append("text").
	attr("text-anchor", "middle").
	attr("x", width/2).
	attr("y", height+margins.bottom/2).
	text("Number of genomes");

svg.append("text").
	attr("text-anchor", "middle").
	attr("x", -margins.left/2).
	attr("y", height/2).
	attr("transform", "rotate(-90," + -margins.left/2+ ","+height/2+")").
	text("Number of {{type_txt}}");

svg.append("g").
	attr("transform", "translate(0,"+ height +")").
	call(d3.axisBottom(x_scale).ticks(data_count.length-1, "d"));

svg.append("g").
	call(d3.axisLeft(y_scale));

var enumerate = function(d) { 
	let arr = [];
	for(var i=0; i<d.length; i++) {
		arr.push([i+1, d[i]]);
	}
	return arr;
}


svg.append("path").
	datum(enumerate(data_acc)).
	attr("fill", "none").
	attr("stroke", acc_color).
	attr("stroke-width", 1.5).
	attr("d", d3.line()
		.x(function(d) { return x_scale(d[0]); })
		.y(function(d) { return y_scale(d[1]); })
	);

svg.append("path").
	datum(enumerate(data_count)).
	attr("fill", "none").
	attr("stroke", shared_color).
	attr("stroke-width", 1.5).
	attr("d", d3.line()
		.x(function(d) { return x_scale(d[0]); })
		.y(function(d) { return y_scale(d[1]); })
	);

svg.append("g").
	selectAll("dot").
	data(enumerate(data_acc)).
	enter().
	append("circle").
	attr("cx", function(d) { return x_scale(d[0]); }).
	attr("cy", function(d) { return y_scale(d[1]); }).
	attr("r", 5).
	attr("fill", acc_color);

svg.append("g").
	selectAll("dot").
	data(enumerate(data_count)).
	enter().
	append("circle").
	attr("cx", function(d) { return x_scale(d[0]); }).
	attr("cy", function(d) { return y_scale(d[1]); }).
	attr("r", 5).
	attr("fill", shared_color);
</script>

</html>
